include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,209 @@
+     */
+    pnode();
+
+    /**
+     * Constructor for bool
+     *
+     * @param boolean  the bool type value
+     */
+    pnode(bool boolean);
+
+    /**
+     * Constructor for integer with all integer type
+     *
+     * @tparam T  input type
+     *
+     * @param integer  the integer type value
+     */
+    template <typename T,
+              std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    pnode(T integer);
+
+    /**
+     * Constructor for string
+     *
+     * @param str  string type value
+     */
+    pnode(const std::string& str);
+
+    /**
+     * Constructor for char* (otherwise, it will use bool)
+     *
+     * @param str  the string like "..."
+     */
+    pnode(const char* str);
+
+    /**
+     * Constructor for double
+     *
+     * @param real  the floating point type value
+     */
+    pnode(double real);
+
+    /**
+     * Constructor for array
+     *
+     * @param array  an pnode array
+     */
+    pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node status.
+     *
+     * @return the status
+     */
+    status_t get_status() const;
+
+    /**
+     * Get the array. It will throw error if the current node does not hold an
+     * array.
+     *
+     * @return the array const reference
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Get the map. It will throw error if the current node does not hold an
+     * map.
+     *
+     * @return the map const reference
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Get the boolean value. It will throw error if the current node does not
+     * hold an boolean value.
+     *
+     * @return the boolean value
+     */
+    bool get_boolean() const;
+
+    /**
+     * Get the integer value. It will throw error if the current node does not
+     * hold an integer value.
+     *
+     * @return the integer value with type int64_t
+     */
+    std::int64_t get_integer() const;
+
+    /**
+     * Get the real floating point value. It will throw error if the current
+     * node does not hold an real value.
+     *
+     * @return the real value with type double
+     */
+    double get_real() const;
+
+    /**
+     * Get the string. It will throw error if the current node does not hold an
+     * string.
+     *
+     * @return the string
+     */
+    std::string get_string() const;
+
+    /**
+     * This function is to access the data under the map. It will throw error
+     * when it does not hold a map. When access non-existent path in the map, it
+     * will return an empty node.
+     *
+     * @param path  the key of the map
+     *
+     * @return node const reference. If the map does not have the path, return
+     * an empty node.
```

Comment:
I know I brougt the empty node as default return value up, but I could also imagine to have to versions, one with a single argument which throws for invalid access, and a two argument version with a default value. 

For me it's fine as it is, just in case others want to change this,

Reply:
I think it can be one with bool, like this?
```
auto get(key, bool throw_error = false) {
  // not-found
  if (throw_error) {throw;}
  else {return empty?}
}
```

Reply:
I'm not sure how much value this would add. A two parameter version would allow us to use default values for our factories quite easily.

Reply:
like this?
```
factory.with_reduction_factor(get_value<type>(config.get("reduction_factor", factory.reduction_factor)));
```
I am not sure. it may be useful in integer or real value, but other complex or pointer will be a bit long code

Reply:
Yes, but I guess it's not important.